client/dfclient/src/websocket.cpp

```diff
@@ -90,79 +90,103 @@ bool WebSocketEmscripten::Send(std::string& data) {
 public:
 	int Connect(std::string& address) override;
 	bool Send(std::string& data) override;
-	virtual ~WebSocketPP() {}
+	virtual ~WebSocketBeast() {}
 
-	WebSocketPPClient c;
-	websocketpp::connection_hdl myHdl;
 };
 
-typedef WebSocketPP WebSocketType;
+typedef WebSocketBeast WebSocketType;
 
-std::map<void*, WebSocketPP*> clients;
-
-void WebSocketPPOnMessage(websocketpp::connection_hdl hdl, WebSocketPPClient::message_ptr msg) {
-	WebSocketPP* wspp = clients[hdl.lock().get()];
-	std::lock_guard<std::mutex> guard(wspp->mq_mutex);
-	wspp->messageQueue.push_back(msg->get_payload());
-}
+#include <boost/beast/core.hpp>
+#include <boost/beast/websocket.hpp>
+#include <boost/asio/connect.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <cstdlib>
+#include <iostream>
+#include <string>
+#include <thread>
 
-void WebSocketPPOnOpen(websocketpp::connection_hdl hdl) {
-	WebSocketPP* wspp = clients[hdl.lock().get()];
-	std::lock_guard<std::mutex> guard(wspp->mq_mutex);
-	wspp->toBeOpened = true;
-}
+namespace beast = boost::beast;         // from <boost/beast.hpp>
+namespace http = beast::http;           // from <boost/beast/http.hpp>
+namespace websocket = beast::websocket; // from <boost/beast/websocket.hpp>
+namespace net = boost::asio;            // from <boost/asio.hpp>
+using tcp = boost::asio::ip::tcp;       // from <boost/asio/ip/tcp.hpp>
 
-int WebSocketPP::Connect(std::string& address) {
-	// Set logging to be pretty verbose (everything except message payloads)
-	c.set_access_channels(websocketpp::log::alevel::none);
-	c.set_error_channels(websocketpp::log::elevel::all);
+net::io_context ioc;
+websocket::stream<tcp::socket> ws{ioc};
+beast::flat_buffer buffer;
 
-	// Initialize ASIO
-	c.init_asio();
+void DoRead();
 
-	// Register our message handler
-	c.set_message_handler(&WebSocketPPOnMessage);
-	c.set_open_handler(&WebSocketPPOnOpen);
+void OnRead(beast::error_code ec, std::size_t bytes_transferred) {
+	if (ec == websocket::error::closed)
+		return;
 
-	websocketpp::lib::error_code ec;
-	WebSocketPPClient::connection_ptr con = c.get_connection(address, ec);
 	if (ec) {
-		std::cout << "could not create connection because: " << ec.message() << std::endl;
-		return -1;
+		if (ec.value() == boost::system::errc::operation_canceled) {
+			// this websocket has disconnected
+			// todo: cleanup whatever is happening and return to main menu
+			return;
+		}
+		std::cout << "read failed " << ec << "\n";
 	}
 
-	// Note that connect here only requests a connection. No network messages are
-	// exchanged until the event loop starts running in the next line.
-	c.connect(con);
-	this->myHdl = con;
+	auto str = beast::buffers_to_string(buffer.data());
+	WebSocketBeast* wsb = (WebSocketBeast*) webSocketManager._ws.get();
+	std::lock_guard<std::mutex> guard(wsb->mq_mutex);
+	wsb->messageQueue.push_back(str);
+	buffer.consume(buffer.size());
+	DoRead();
+}
 
-	clients[this->myHdl.lock().get()] = this;
+void DoRead() {
+	ws.async_read(buffer, &OnRead);
```

Comment:
Am I getting something wrong, or is it infinite recursion? Is it the official way to do this, isn't it something to avoid?

Reply:
This just sets up an asynchronous operation. The call returns immediately and OnRead will be called back when some data will be available to read.